### ISSUE TITLE
Dynamic asynchronous context menu

### ISSUE THREAD

**Issue by mathpaquette** [ag-grid/ag-grid#3518]
**I'm submitting an idea/Enhancement **

[ X] I want dynamic async context menus, which make api call and return menus. 
It is not possible to do with existing functionalities. 
Current behavior
On right click of the Ag-Grid we are able to get context menu, but unable to have dynamic menu items that are based on row node's data[Data got through API call].

Expected behavior
Dynamic menus should be possible, add a prop in list and it should be considered as dynamic async menu list, it should give a loading indicator till it fetches the items in API call.

Minimal reproduction of the problem with instructions
https://next.plnkr.co/edit/lMqw5MZSUrmqSToj

What is the motivation / use case for changing the behavior?
User will be able to easy perform actions, instead of some other workflow implementation.

Please tell us about your environment:

ag-Grid-Enterprise version: 19.0.0

Browser: [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
Language: [all | TypeScript X.X | ES6/7 | ES5]

**Comment by mathpaquette** [comment]
took it from @vijayanandof  https://github.com/ag-grid/ag-grid-enterprise/issues/223

**Comment by mathpaquette** [comment]
any clean workaround for doing this?

**Comment by icedjello** [comment]
Hi

Our consultants have provided the following example for you:
https://next.plnkr.co/edit/zUtHvyaqv0ORd0Q8

Perhaps you could leverage this approach. You could also use a debounce to ensure only one context menu is open at a time.

We will flag this issue as managed by community as ag-Grid do not provide support over GitHub.

Ahmed‍

**Comment by mathpaquette** [comment]
> Hi
> 
> Our consultants have provided the following example for you:
> https://next.plnkr.co/edit/zUtHvyaqv0ORd0Q8
> 
> Perhaps you could leverage this approach. You could also use a debounce to ensure only one context menu is open at a time.
> 
> We will flag this issue as managed by community as ag-Grid do not provide support over GitHub.
> 
> Ahmed‍

awesome thank you very much, ill review it.
Mathieu

**Comment by mathpaquette** [comment]
@icedjello tried your example on plunker and it doesn't show anything. it is a showcase for background code only? thanks.

**Comment by vijayanandof** [comment]
yeah,@mathpaquette. I think @icedjello is suggesting a approach for code.

**Comment by tanjianyong** [comment]
@icedjello   so we can  this api( params.api.contextMenuFactory.showMenu(params.node, params.column, params.value, params.event)) to refresh the context menu that opened? I tried but  the context menu didn't refresh.

@mathpaquette  I also get the problem, have you solved that? Appreciate it if you have any idea to share.

**Comment by mathpaquette** [comment]
@tanjianyong unfortunately I raised the issue for one member in my team but I can check the workaround they used to solve this. Which version and which framework are you using?
Cheers.

**Comment by tanjianyong** [comment]
@mathpaquette  thanks for your reply.
My application is based on  react(16.0.0), dva(2.1.0) and ag-grid-enterprise(18.0.0).
The problem I get is the context menu should be based on the cell that user clicked. It seems that context menu can't be based on the remote api data.

**Comment by tanjianyong** [comment]
@mathpaquette  thanks a lot.I've fixed the problem just by setting setTimeout, but I still want know if any better solution to wait for the request reponse.

**Comment by Bhagwan96** [comment]
@tanjianyong @mathpaquette If this issue is fixed, can any of you please share the code snippet how to fetch context menu data asynchronously via API call and show it it context menu ? React code would be appreciated :) . Thanks

**Comment by tanjianyong** [comment]
Hi @Bhagwan96, of course I can share. 
In react I use fetch to request and it returns the Promise object. Then I  assemble the object the ag-grid context menu needs(getContextMenuItems) in models(the part of processing state in Dva  ) . And I use the (params.api.contextMenuFactory.showMenu(params.node, params.column, params.value, params.event) api to refresh the context menu. Now it works.

**Comment by Bhagwan96** [comment]
@tanjianyong I didn't find showMenu function in contextMenuFactory object. Could you please attach the code snippet for getContextMenuItems(paramas) like how to use gridApi to refresh context menu ?

**Comment by mathpaquette** [comment]
@Bhagwan96 did you solve it?

**Comment by Bhagwan96** [comment]
Yeah It's working fine, but I could not find a way to show loading indicator until fetch completed(Loading indicator in context Menu).

**Comment by sukhminderx** [comment]
Hello
I'm not able to find the code in getContextMenuItems(params). Mine code is displaying the snyc menu instead of the one returned by the promise.
Can someone please help with that.

**Comment by salamaashoush** [comment]
@sukhminderx @mathpaquette  here is how I fixed this, add these to grid config

```
getContextMenuItems: () => [],
onCellContextMenu: params => {
  this.getMenuItemsFromSomewhereAsync(params).then(menuItems => {
          const factory = params.api.contextMenuFactory; // if you use ts cast api to `any` becasue factory is private 
          factory.getMenuItems = () => menuItems; // must override this method otherwise wont work
          factory.showMenu(params.node, params.column, params.value, params.event);
   });
  }
```

**Comment by shivaaudam** [comment]
@tanjianyong for me contextMenuFactory is null,  Could you please attach the code snippet how it is working?

**Comment by feyzo** [comment]
If you have used @salamaashoush's suggestion consider making the following change. We found a bug when we added another menu and it always displayed the custom menu (`menuItems`). This was because we were overwriting the factory to always display the custom menu in below code. Solution was to switch back to ag-grid's own function after displaying the custom menu.

```
const factory = params.api.contextMenuFactory;

// Cache ag-grid's own function so we can reassign it
const defaultGetMenuItems = factory.getMenuItems;

factory.getMenuItems = () => menuItems;
factory.showMenu(params.node, params.column, params.value, params.event);

// Revert the overwritten function back to ag-grid's own function
// Without this reversion any menu feature will open the `menuItems`
// because we have overritten ag-grid's menu handler function to always display our custom menu
factory.getMenuItems = defaultGetMenuItems;
```

**Comment by patriceo** [comment]
Hi I'm having the same issue, probably it would be convenient to augment the existing API with an Observable option such like:

```
export interface GetContextMenuItems {
    (params: GetContextMenuItemsParams): (string | MenuItemDef)[]| Observable<(string | MenuItemDef)[]>;
}

```

I'm scared that using the private contextMenuFactory of the API is quite risky if for any reason the internal implementation changes.

**Comment by JeffreyZhao** [comment]
That's how I patched ag-grid to support async loading of context menu. Now the `getContextMenuItems` can return a Promise which will be resolved to show the context menu delayed.

```ts
import { GridBodyController } from 'ag-grid-community';

const patched = Symbol('Patched') as unknown as string;
const origAddBodyViewportListener = GridBodyController.prototype['addBodyViewportListener'];

// eslint-disable-next-line @typescript-eslint/no-explicit-any
GridBodyController.prototype['addBodyViewportListener'] = function (this: any, ...args: any[]) {
  if (!this.contextMenuFactory[patched]) {
    if (typeof jest === 'undefined') {
      console.log('=== Patching AgGrid: ContextMenuFactory');
    }

    const origGetMenuItems = this.contextMenuFactory.getMenuItems;

    let loadedMenuItems;
    this.contextMenuFactory.getMenuItems = function (...args: any[]) {
      const items = loadedMenuItems;
      loadedMenuItems = undefined;

      if (items) {
        return items;
      }

      return origGetMenuItems.apply(this, args);
    };

    let latestLoadingId = 0;

    const origShowMenu = this.contextMenuFactory.showMenu;
    this.contextMenuFactory.showMenu = function (node, column, value, mouseEvent, anchorToElement) {
      const itemsOrPromise = this.getMenuItems(node, column, value);
      if (!itemsOrPromise) {
        return false;
      }

      const loadingId = ++latestLoadingId;

      if (itemsOrPromise instanceof Promise) {
        itemsOrPromise.then(items => {
          if (loadingId !== latestLoadingId) {
            return;
          }

          loadedMenuItems = items;
          origShowMenu.call(this, node, column, value, mouseEvent, anchorToElement);
        });
      } else {
        loadedMenuItems = itemsOrPromise;
        origShowMenu.call(this, node, column, value, mouseEvent, anchorToElement);
      }

      return true;
    };

    this.contextMenuFactory[patched] = true;
  }

  return origAddBodyViewportListener.apply(this, args);
};
```

**Comment by IgorPropisnov** [comment]
> 

How and where to insert this patch?

**Comment by IgorPropisnov** [comment]
> That's how I patched ag-grid to support async loading of context menu. Now the `getContextMenuItems` can return a Promise which will be resolved to show the context menu delayed.
> 
> ```ts
> import { GridBodyController } from 'ag-grid-community';
> 
> const patched = Symbol('Patched') as unknown as string;
> const origAddBodyViewportListener = GridBodyController.prototype['addBodyViewportListener'];
> 
> // eslint-disable-next-line @typescript-eslint/no-explicit-any
> GridBodyController.prototype['addBodyViewportListener'] = function (this: any, ...args: any[]) {
>   if (!this.contextMenuFactory[patched]) {
>     if (typeof jest === 'undefined') {
>       console.log('=== Patching AgGrid: ContextMenuFactory');
>     }
> 
>     const origGetMenuItems = this.contextMenuFactory.getMenuItems;
> 
>     let loadedMenuItems;
>     this.contextMenuFactory.getMenuItems = function (...args: any[]) {
>       const items = loadedMenuItems;
>       loadedMenuItems = undefined;
> 
>       if (items) {
>         return items;
>       }
> 
>       return origGetMenuItems.apply(this, args);
>     };
> 
>     let latestLoadingId = 0;
> 
>     const origShowMenu = this.contextMenuFactory.showMenu;
>     this.contextMenuFactory.showMenu = function (node, column, value, mouseEvent, anchorToElement) {
>       const itemsOrPromise = this.getMenuItems(node, column, value);
>       if (!itemsOrPromise) {
>         return false;
>       }
> 
>       const loadingId = ++latestLoadingId;
> 
>       if (itemsOrPromise instanceof Promise) {
>         itemsOrPromise.then(items => {
>           if (loadingId !== latestLoadingId) {
>             return;
>           }
> 
>           loadedMenuItems = items;
>           origShowMenu.call(this, node, column, value, mouseEvent, anchorToElement);
>         });
>       } else {
>         loadedMenuItems = itemsOrPromise;
>         origShowMenu.call(this, node, column, value, mouseEvent, anchorToElement);
>       }
> 
>       return true;
>     };
> 
>     this.contextMenuFactory[patched] = true;
>   }
> 
>   return origAddBodyViewportListener.apply(this, args);
> };
> ```
How and where to insert this patch?

**Comment by vardhan03** [comment]
@patriceo @mathpaquette @shivaaudam 
We have this requirement for one of our ongoing projects. 
I am scared to use the private API method 'contextMenuFactory' of the grid.
Is there any other preferable way?

**Comment by Jrubzjeknf** [comment]
`contextMenuFactory` isn't available any more on the grid api since v32. I found that the easiest way to get a hold of it is to get it from a column node. Use the event `columnEverythingChanged` with this function:

```ts
	handleColumnEverythingChanged(event: FqsGridEvent): void {
		if (this.agGridInternalsCollection) return;

		const column = this.gridApi?.api?.getColumns()?.[0] as any;

		if (!column) return;

		this.agGridInternalsCollection = column.stubContext.beans as CoreBeanCollection;
	}
```

Beans is ag grid's internal dependency injection framework. Be aware this code can break in the next release.

**Comment by Jrubzjeknf** [comment]
In other news, I wrote to the enterprise support and their reply was positive.

>Currently there's no way to implement this behavior with AG Grid. However, we agree this would be a valuable addition to the product and we have added a feature request for this.
> 
>We have added this requirement to our backlog and we are tracking it with the following reference and description:
AG-2613 [Context Menu] Allow context menu items to be provided asynchronously
 >
>This item is currently IN PROGRESS and will be a part of our next major release in early December. 
>
>You can get an update on the status of this item on the product pipeline page:
https://ag-grid.com/pipeline/

Looks like we won't have to use hacks for long any more. :)

**Comment by vardhan03** [comment]
Thank you for the update @Jrubzjeknf, yes, no more hacks !!
We have reached out to enterprise support and got the same reply.
I am looking forward to the release.

**Comment by RuslanZavacky** [comment]
@Jrubzjeknf just to clarify, it seems that 2613 was moved to Backlog. Is it going to be in next release?

**Comment by Jrubzjeknf** [comment]
@RuslanZavacky Well, I guess they reprioritized and this didn't make the cut. Hopefully it'll come with the next one then.

**Comment by Jrubzjeknf** [comment]
@vardhan03 @RuslanZavacky Async context menu has been released in 33.0.0. `getContextMenuItems` is now of type `(params: GetContextMenuItemsParams<TData, TContext>) : MenuCallbackReturn<DefaultMenuItem, TData, TContext> | Promise<MenuCallbackReturn<DefaultMenuItem, TData, TContext>>`.
